include/object/schedcontext.h

```diff
@@ -66,6 +66,10 @@ void schedContext_bindNtfn(sched_context_t *sc, notification_t *ntfn);
 /* unbind scheduling context from a notification */
 void schedContext_unbindNtfn(sched_context_t *sc);
 
+/* Return a donated scheduling context to the notification to which it
```

Comment:
If I am reading the commit message correctly, this changes the semantics such that *any* blocking will return a notification SC to the notification?

The semantics I intended (even if I didfn't write it down correctly) were that blocking on the notification, or blocking on an endpoint when you are bound (anything that *ends the notification operation normally*) the SC would go back.  

What should not happen is that an unrelated blocking operation causes the SC to return. I need to be able to do blocking I/O via other notifications. Does this still work with this change?

Reply:
It does not, this was an overzealous change to resolve an issue where, after receiving an SC from a bound notification, the only way to return it was to recv directly on the bound notification. Kent has suggested a refinement where rather than returning on any IPC it returns on any receive, but this may still be too broad.

Reply:
My suggestion was the same as @pingerino, a receive on the notification object or any endpoint if the notification is bound.  Although this could still be too restrictive - it doesn't seem that the kernel can unambiguously know when the thread means to _end the notification operation normally_.  For example a server that wishes to rate limit the amount of messages it receives could wish to mostly block on the notification object for a periodic timer notification and then perform a seL4_Recv to get a message won't be able to ensure that it's SC doesn't go back to the notification.  It seems that returning the donated SC to the notification object should be a similar process to that of one donated from an endpoint - a reply object and a NBSendWait or NBSendRecv operation...

Reply:
Currently, sending on an endpoint doesn't donate any SC. It would make sense for a send on a notification to return the SC in some (but not all) cases.